icicle/appUtils/large_ntt/kernel_ntt.cu

```diff
@@ -0,0 +1,640 @@
+    s_meta.ntt_inp_id = strided ? (threadIdx.x >> 5) : (threadIdx.x & 0x1);
+
+    engine.loadBasicTwiddles(basic_twiddles, inv);
+    engine.loadGlobalData(in, data_stride, log_data_stride, log_size, strided, s_meta);
+    engine.loadInternalTwiddles16(internal_twiddles, strided, inv);
+    engine.ntt8win();
+    engine.twiddlesInternal();
+    engine.SharedData16Columns8(shmem, true, false, strided); // store
+    __syncthreads();
+    engine.SharedData16Rows2_4(shmem, false, false, strided); // load
+    engine.ntt2_4();
+    if (twiddle_stride) {
+      engine.loadExternalTwiddlesGeneric16(
+        external_twiddles, twiddle_stride, log_data_stride, s_meta, tw_log_size, inv);
+      engine.twiddlesExternal();
+    }
+    engine.storeGlobalData16(out, data_stride, log_data_stride, log_size, strided, s_meta);
+  }
+
+  template <typename E, typename S>
+  __launch_bounds__(64) __global__ void ntt16dit(
+    E* in,
+    E* out,
+    S* external_twiddles,
+    S* internal_twiddles,
+    S* basic_twiddles,
+    uint32_t log_size,
+    uint32_t tw_log_size,
+    uint32_t data_stride,
+    uint32_t log_data_stride,
+    uint32_t twiddle_stride,
+    bool strided,
+    uint32_t stage_num,
+    bool inv,
+    bool dit)
+  {
+    NTTEngine<E, S> engine;
+    stage_metadata s_meta;
+
+    SharedMemory<E> smem;
+    E* shmem = smem.getPointer();
+
+    s_meta.th_stride = 2;
+    s_meta.ntt_block_size = 16;
+    s_meta.ntt_block_id = (blockIdx.x << 5) + (strided ? (threadIdx.x & 0x1f) : (threadIdx.x >> 1));
+    s_meta.ntt_inp_id = strided ? (threadIdx.x >> 5) : (threadIdx.x & 0x1);
+
+    engine.loadBasicTwiddles(basic_twiddles, inv);
+    engine.loadGlobalData16(in, data_stride, log_data_stride, log_size, strided, s_meta);
+    if (twiddle_stride) {
+      engine.loadExternalTwiddlesGeneric16(
+        external_twiddles, twiddle_stride, log_data_stride, s_meta, tw_log_size, inv);
+      engine.twiddlesExternal();
+    }
+    engine.loadInternalTwiddles16(internal_twiddles, strided, inv);
+    engine.ntt2_4();
+    engine.SharedData16Columns2_4(shmem, true, false, strided); // store
+    __syncthreads();
+    engine.SharedData16Rows8(shmem, false, false, strided); // load
+    engine.twiddlesInternal();
+    engine.ntt8win();
+    engine.storeGlobalData(out, data_stride, log_data_stride, log_size, strided, s_meta);
+  }
+
+  template <typename E, typename S>
+  __global__ void normalize_kernel(E* data, S norm_factor)
+  {
+    uint32_t tid = blockIdx.x * blockDim.x + threadIdx.x;
+    data[tid] = data[tid] * norm_factor;
+  }
+
+  template <typename S>
+  __global__ void generate_base_table(S basic_root, S* base_table, uint32_t skip)
+  {
+    S w = basic_root;
+    S t = S::one();
+    for (int i = 0; i < 64; i += skip) {
+      base_table[i] = t;
+      t = t * w;
+    }
+  }
+
+  template <typename S>
+  __global__ void generate_basic_twiddles(S basic_root, S* w6_table, S* basic_twiddles)
```

Comment:
Just fyi: I tried several alternative ways to store basic twiddles like GPU const memory and even hardcoding them in but nothing gave any speedups